### ISSUE TITLE
PSM and PSM2 symbol conflicts

### ISSUE THREAD

**Issue by jsquyres** [ofiwg/libfabric#1258]
@j-xiong Because of an issue in Open MPI regarding PSM vs. PSM2, it just dawned on me what sub-bullet 4 in the commit message for a076ab33b7b28580db37b5c01aeddef7f0805248 means:

> (4) at run time, only one of the two providers (PSM vs. PSM2) can succeed in initialization, depending on the actual PSM library in use.

Please correct me if I'm wrong, but the issue is that `libpsm_infinipath` and `libpsm2` both declare the same symbols.  For example, they both define the `psm_init()` symbol.

How does that work at linking time?  E.g., if `libfabric` is linked against both `libpsm_infinipath` and `libpsm2`, how do we not get symbol conflicts?  Is it because the run time linker will simply resolve the required symbols with whichever library -- psm1 or psm2 -- just happens to be loaded first?  And if that's the case, how is it good that the linker loading order gets to decide whether you get to use PSM1 or PSM2?

Just curious: why doesn't PSM2 have different symbols than PSM1?  Having PSM1 and PSM2 have the same symbols -- potentially with different meanings? -- sounds like we're entering libnl-vs-libnl3 territory (which is a terrible, terrible place to be...).

Am I misunderstanding the situation?  Is there some other magic that is happening that makes this all work out ok?

**Comment by j-xiong** [comment]
@jsquyres I agree that the message doesn't capture the full picture. It can work under certain conditions, mostly when LD_PRELOAD is used to preload the intended library.  But in general it's problematic.

I am working on a patch to handle the issue of PSM1/PSM2 coexistence. Hopefully the PR will be created in one or two days.

PSM2 is a super set of PSM1, it's backward compatible in terms of API. Unfortunately the lower level is not compatible with TrueScale.

**Comment by jsquyres** [comment]
But that's the exact problem: PSM2 is *not* a superset of PSM1.  My understanding -- and please correct me if I'm wrong -- is:

* PSM1: contains symbols in set A, and support network type X
* PSM2: contains symbols in set A and set B, and supports network type Y

If PSM2 were a superset of PSM1, then it would contain symbols A and B *and* support **BOTH** network types X and Y.  But this is not the case.

As such, PSM1 and PSM2 are effectively unrelated libraries who happen to share many of the same symbols.  This makes it very, very hard to support both PSM1 and PSM2 in the same software package: you're forcing both Open MPI and libfabric to have both complicated configury *and* forcing them to both use an indirection layer to get to the PSM1/PSM2 functionality (i.e., use dlopen and function pointers).  It seems like you'll probably have this same issue -- and same convoluted solution -- in any software that uses PSM1/PSM2.  That sounds very undesirable, IMHO.

FWIW: this is not a common upgrade strategy for libraries when evolving to a new generation of their software.  Typically, there will be either an entirely new library that does not share any symbols with the old library, or the next-gen library is an actual superset of the old library.

Specifically: if PSM2 were an actual superset of PSM1, then using PSM-based networks *significantly* easier for software developers: if PSM2 is available, use it.  Otherwise, use PSM1.  That's simple and easy to explain and understand.

In short: if you have to have extremely confusing conversations in multiple software communities (e.g., http://www.open-mpi.org/community/lists/devel/2015/09/17919.php) explaining your configury, function-pointer-indirection code (i.e., PR #1259), compilation, and linking scheme to upgrade to a new library, you're doing it wrong.

**Comment by j-xiong** [comment]
@jsquyres I agree with you on what you said about PSM1 and PSM2. The "super set" term I used only referred to the API, not the library. I would also like to see that the PSM2 library works for both networks but that is beyond my control.

**Comment by jsquyres** [comment]
Just following up on this issue: we got a response from the PSM maintainers on the Open MPI side (from @afriedle-intel): http://www.open-mpi.org/community/lists/devel/2015/09/17953.php.

Hopefully it means that this issue can be closed when the PSM maintainers fix the issue internal to the PSM2 library.
